code/modules/power/gravity_generator.dm

```diff
@@ -0,0 +1,571 @@
+	name = "gravitational generator"
+	desc = "A device which produces a graviton field when set up."
+	icon = 'icons/obj/machines/gravity_generator.dmi'
+	anchored = TRUE
+	density = TRUE
+	use_power = FALSE
+	unacidable = TRUE
+
+	light_color = "#7de1e1"
+
+	var/sprite_number = 0
+	var/broken_state = 0
+
+/obj/machinery/gravity_generator/on_update_icon()
+	icon_state = "[broken_state]_[sprite_number]"
+
+/obj/machinery/gravity_generator/proc/show_broken_info()
+	return
+
+/obj/machinery/gravity_generator/explosion_act(severity)
+	return
+
+/obj/machinery/gravity_generator/emp_act(severity)
+	return
+
+/obj/machinery/gravity_generator/bullet_act(obj/item/projectile/P, def_zone)
+	return
+
+/obj/machinery/gravity_generator/take_damage(amount)
+	return
+
+//
+// Generator which spawns with the station.
+//
+
+GLOBAL_VAR(station_gravity_generator)
+/obj/machinery/gravity_generator/main/station/Initialize(mapload, ...)
+	. = ..()
+	GLOB.station_gravity_generator = src
+	if(!mapload)
+		update_connectected_areas_gravity()
+
+
+/obj/machinery/gravity_generator/main/station/Destroy()
+	if(GLOB.station_gravity_generator == src)
+		GLOB.station_gravity_generator = null
+	return ..()
+
+//
+// Main Generator with the main code
+//
+
+/obj/machinery/gravity_generator/main
+	name = "gravitational generator panel"
+	icon_state = "0_8"
+	idle_power_usage = 0
+	active_power_usage = 100000
+	power_channel = ENVIRON
+	sprite_number = 8
+	use_power = POWER_USE_ACTIVE
+
+	var/enabled = TRUE               // for switching gravity status in areas
+	var/breaker = TRUE               // if true - charges the GG if it has power
+	var/charging_state = POWER_IDLE  // check process()
+	var/charge_count = 100           // % of current charge
+	var/health = 1000
+
+	var/list/parts = list()
+	var/list/lights = list()
+	var/list/connected_areas = list()
+	var/obj/machinery/gravity_generator/part/middle = null
+
+	// Wires
+	var/announcer = TRUE                  // if true - notifies about the switching of the state of the generator to the engineering channel
+	var/power_supply = TRUE               // if false - will lose power after proc/update_power()
+	var/can_toggle_breaker = TRUE
+	var/emergency_shutoff_button = FALSE  // if true - shows an additional option with emergency generator shutdown
+
+/obj/machinery/gravity_generator/main/Initialize()
+	. = ..()
+	setup_parts()
+	update_icon()
+	add_areas()
+	wires = new(src)
+
+/obj/machinery/gravity_generator/main/Destroy()
+	QDEL_NULL(wires)
+	for(var/obj/machinery/gravity_generator/part/P in parts)
+		P.main_part = null
+		parts -= P
+		if(!QDELETED(P))
+			qdel(P)
+	middle = null
+	lights = null
+	connected_areas = null
+	update_connectected_areas_gravity()
+	return ..()
+
+/obj/machinery/gravity_generator/main/examine(mob/user)
+	. = ..()
+	if(panel_open)
+		to_chat(user, "The maintenance hatch is open.")
+	show_broken_info(user)
+
+/obj/machinery/gravity_generator/main/show_broken_info(mob/user)
```

Comment:
Might be worth making repair mechanics part of the machine construction datum instead of a hardcoded thing.